File: src/nbody.cpp
#include "nbody.h"
#include <cmath>
#include <iostream>
#include <iomanip>

NBodySimulation::NBodySimulation(const std::vector<double> &masses_vec,
                                 const std::vector<std::vector<double>> &positions_vec,
                                 const std::vector<std::vector<double>> &velocities_vec,
                                 double G_val)
    : G(G_val), n_bodies(masses_vec.size())
{

    if (positions_vec.size() != n_bodies || velocities_vec.size() != n_bodies)
    {
        throw std::invalid_argument("Size of masses array is " + std::to_string(masses_vec.size()) +
                                    ", positions array size is " + std::to_string(positions_vec.size()) +
                                    " and velocities array size is " + std::to_string(velocities_vec.size()));
    }

    std::cout << VectorXd(n_bodies).size() << " bodies initialized with G = " << G << std::endl;

    masses = VectorXd(n_bodies);
    for (int i = 0; i < n_bodies; ++i)
    {
        masses(i) = masses_vec[i];
    }

    positions = MatrixXd(n_bodies, 3);
    for (int i = 0; i < n_bodies; ++i)
    {
        for (int j = 0; j < 3; ++j)
        {
            positions(i, j) = positions_vec[i][j];
        }
    }

    velocities = MatrixXd(n_bodies, 3);
    for (int i = 0; i < n_bodies; ++i)
    {
        for (int j = 0; j < 3; ++j)
        {
            velocities(i, j) = velocities_vec[i][j];
        }
    }

    history.push_back(positions);
}

MatrixXd NBodySimulation::compute_forces()
{
    MatrixXd forces = MatrixXd::Zero(n_bodies, 3);

    for (int i = 0; i < n_bodies; ++i)
    {
        for (int j = i + 1; j < n_bodies; ++j)
        {
            Vector3d r_vec = positions.row(j) - positions.row(i);
            double r_mag_sq = r_vec.squaredNorm();

            if (r_mag_sq < 1e-20)
                continue;

            double r_mag = std::sqrt(r_mag_sq);
            double force_mag = G * masses(i) * masses(j) / r_mag_sq;
            Vector3d force_vec = (force_mag / r_mag) * r_vec;

            forces.row(i) += force_vec.transpose();
            forces.row(j) -= force_vec.transpose();
        }
    }

    return forces;
}

void NBodySimulation::step_rk4(double dt)
{
    MatrixXd pos0 = positions;
    MatrixXd vel0 = velocities;

    MatrixXd forces1 = compute_forces();
    MatrixXd acc1(n_bodies, 3);
    for (int i = 0; i < n_bodies; ++i)
    {
        acc1.row(i) = forces1.row(i) / masses(i);
    }
    MatrixXd dpos1 = vel0;
    MatrixXd dvel1 = acc1;

    positions = pos0 + 0.5 * dt * dpos1;
    velocities = vel0 + 0.5 * dt * dvel1;
    MatrixXd forces2 = compute_forces();
    MatrixXd acc2(n_bodies, 3);
    for (int i = 0; i < n_bodies; ++i)
    {
        acc2.row(i) = forces2.row(i) / masses(i);
    }
    MatrixXd dpos2 = velocities;
    MatrixXd dvel2 = acc2;

    positions = pos0 + 0.5 * dt * dpos2;
    velocities = vel0 + 0.5 * dt * dvel2;
    MatrixXd forces3 = compute_forces();
    MatrixXd acc3(n_bodies, 3);
    for (int i = 0; i < n_bodies; ++i)
    {
        acc3.row(i) = forces3.row(i) / masses(i);
    }
    MatrixXd dpos3 = velocities;
    MatrixXd dvel3 = acc3;

    positions = pos0 + dt * dpos3;
    velocities = vel0 + dt * dvel3;
    MatrixXd forces4 = compute_forces();
    MatrixXd acc4(n_bodies, 3);
    for (int i = 0; i < n_bodies; ++i)
    {
        acc4.row(i) = forces4.row(i) / masses(i);
    }
    MatrixXd dpos4 = velocities;
    MatrixXd dvel4 = acc4;

    positions = pos0 + dt / 6.0 * (dpos1 + 2 * dpos2 + 2 * dpos3 + dpos4);
    velocities = vel0 + dt / 6.0 * (dvel1 + 2 * dvel2 + 2 * dvel3 + dvel4);

    history.push_back(positions);
}

double NBodySimulation::total_energy()
{
    double ke = 0.0;
    for (int i = 0; i < n_bodies; ++i)
    {
        double v_sq = velocities.row(i).squaredNorm();
        ke += 0.5 * masses(i) * v_sq;
    }

    double pe = 0.0;
    for (int i = 0; i < n_bodies; ++i)
    {
        for (int j = i + 1; j < n_bodies; ++j)
        {
            Vector3d r_vec = positions.row(j) - positions.row(i);
            double r_mag = r_vec.norm();
            if (r_mag > 1e-10)
            {
                pe -= G * masses(i) * masses(j) / r_mag;
            }
        }
    }

    return ke + pe;
}

void NBodySimulation::simulate(double t_end, double dt)
{
    double t = 0.0;
    double initial_energy = total_energy();
    time_steps.clear();
    time_steps.push_back(t);

    while (t + dt <= t_end)
    {
        step_rk4(dt);

        t += dt;
        time_steps.push_back(t);
    }
}

std::vector<std::vector<std::vector<double>>> NBodySimulation::get_history() const
{
    std::vector<std::vector<std::vector<double>>> result;
    for (const auto &frame : history)
    {
        std::vector<std::vector<double>> frame_data;
        for (int i = 0; i < frame.rows(); ++i)
        {
            std::vector<double> particle_pos;
            for (int j = 0; j < frame.cols(); ++j)
            {
                particle_pos.push_back(frame(i, j));
            }
            frame_data.push_back(particle_pos);
        }
        result.push_back(frame_data);
    }
    return result;
}

std::vector<std::vector<double>> NBodySimulation::get_positions() const
{
    std::vector<std::vector<double>> result;
    for (int i = 0; i < positions.rows(); ++i)
    {
        std::vector<double> pos;
        for (int j = 0; j < positions.cols(); ++j)
        {
            pos.push_back(positions(i, j));
        }
        result.push_back(pos);
    }
    return result;
}

std::vector<std::vector<double>> NBodySimulation::get_velocities() const
{
    std::vector<std::vector<double>> result;
    for (int i = 0; i < velocities.rows(); ++i)
    {
        std::vector<double> vel;
        for (int j = 0; j < velocities.cols(); ++j)
        {
            vel.push_back(velocities(i, j));
        }
        result.push_back(vel);
    }
    return result;
}

std::vector<double> NBodySimulation::get_masses() const
{
    std::vector<double> result;
    for (int i = 0; i < masses.size(); ++i)
    {
        result.push_back(masses(i));
    }
    return result;
}

std::vector<double> NBodySimulation::get_time_steps() const
{
    return time_steps;
}

int NBodySimulation::get_n_bodies() const { return n_bodies; }
double NBodySimulation::get_G() const { return G; }

File: nbody_jax/nbody.py
from functools import partial
from typing import Iterable
import jax.numpy as jnp
from jax import jit
import numpy as np


class NBodySimulation:
    """
    JAX-based N-body simulation that maintains the same API as the C++ version.
    """

    def __init__(
        self,
        masses: Iterable[float],
        positions: Iterable[Iterable[float]],
        velocities: Iterable[Iterable[float]],
        G: float = 1.0,
    ) -> None:
        """
        Initialize the N-body simulation.

        Parameters
        ----------
        masses : array_like
            Masses of the bodies.
        positions : array_like
            Initial positions of the bodies (shape: (N, 3)).
        velocities : array_like
            Initial velocities of the bodies (shape: (N, 3)).
        G : float, optional
            Gravitational constant. Default is 1.0.
        """
        self.G = G
        self.masses = jnp.array(masses)
        self.positions = jnp.array(positions)
        self.velocities = jnp.array(velocities)
        self.n_bodies = len(masses)

        # Validate input dimensions
        if self.positions.shape != (self.n_bodies, 3):
            raise ValueError(
                f"Positions shape {self.positions.shape} doesn't match expected ({self.n_bodies}, 3)"
            )
        if self.velocities.shape != (self.n_bodies, 3):
            raise ValueError(
                f"Velocities shape {self.velocities.shape} doesn't match expected ({self.n_bodies}, 3)"
            )

        self.history = [self.positions.copy()]
        self.time_steps = []

        self._compute_forces_jit = jit(self._compute_forces_impl)
        self._rk4_step_jit = jit(self._rk4_step_impl)

        print(f"{self.n_bodies} bodies initialized with G = {self.G}")

    @partial(jit, static_argnums=(0,))
    def _compute_forces_impl(self, positions, masses, G):
        delta = positions[None, :, :] - positions[:, None, :]  # (N, N, 3)
        dist_sq = jnp.sum(delta**2, axis=-1) + 1e-9
        dist = jnp.sqrt(dist_sq)
        force_magnitudes = G * masses[:, None] * masses[None, :] / dist_sq
        force_vectors = force_magnitudes[..., None] * delta / dist[..., None]
        net_forces = jnp.sum(
            jnp.where(jnp.eye(len(masses), dtype=bool)[..., None], 0.0, force_vectors),
            axis=1,
        )
        return net_forces

    @partial(jit, static_argnums=(0,))
    def _rk4_step_impl(self, positions, velocities, masses, G, dt):
        forces1 = self._compute_forces_impl(positions, masses, G)
        acc1 = forces1 / masses[:, None]
        dpos1 = velocities
        dvel1 = acc1

        pos2 = positions + 0.5 * dt * dpos1
        vel2 = velocities + 0.5 * dt * dvel1
        forces2 = self._compute_forces_impl(pos2, masses, G)
        acc2 = forces2 / masses[:, None]
        dpos2 = vel2
        dvel2 = acc2

        pos3 = positions + 0.5 * dt * dpos2
        vel3 = velocities + 0.5 * dt * dvel2
        forces3 = self._compute_forces_impl(pos3, masses, G)
        acc3 = forces3 / masses[:, None]
        dpos3 = vel3
        dvel3 = acc3

        pos4 = positions + dt * dpos3
        vel4 = velocities + dt * dvel3
        forces4 = self._compute_forces_impl(pos4, masses, G)
        acc4 = forces4 / masses[:, None]
        dpos4 = vel4
        dvel4 = acc4

        new_positions = positions + dt / 6.0 * (dpos1 + 2 * dpos2 + 2 * dpos3 + dpos4)
        new_velocities = velocities + dt / 6.0 * (dvel1 + 2 * dvel2 + 2 * dvel3 + dvel4)

        return new_positions, new_velocities

    def compute_forces(self):
        """
        Compute current gravitational forces.

        Returns
        -------
        np.ndarray
            Forces array of shape (n_bodies, 3).
        """
        forces_jax = self._compute_forces_jit(self.positions, self.masses, self.G)
        return np.array(forces_jax)

    def step_rk4(self, dt: float) -> None:
        """
        Perform one RK4 integration step.

        Parameters
        ----------
        dt : float
            Time step size.
        """
        self.positions, self.velocities = self._rk4_step_jit(
            self.positions, self.velocities, self.masses, self.G, dt
        )
        self.history.append(self.positions.copy())

    def total_energy(self) -> float:
        """
        Calculate total energy (kinetic + potential) of the system.

        Returns
        -------
        float
            Total energy of the system.
        """
        v_squared = jnp.sum(self.velocities**2, axis=1)
        ke = 0.5 * jnp.sum(self.masses * v_squared)

        r_vectors = self.positions[:, None, :] - self.positions[None, :, :]
        r_distances = jnp.sqrt(jnp.sum(r_vectors**2, axis=-1))

        # Avoid self-interaction and division by zero
        r_distances = jnp.where(r_distances < 1e-10, jnp.inf, r_distances)

        # Upper triangular mask to avoid double counting
        i, j = jnp.triu_indices(self.n_bodies, k=1)
        pe_pairs = -self.G * self.masses[i] * self.masses[j] / r_distances[i, j]
        pe = jnp.sum(pe_pairs)

        return float(ke + pe)

    def simulate(self, t_end: float, dt: float) -> None:
        """
        Run the complete simulation.

        Parameters
        ----------
        t_end : float
            End time of the simulation.
        dt : float
            Time step size.
        """
        t = 0.0
        self.time_steps = [t]

        while t + dt <= t_end:
            self.step_rk4(dt)
            t += dt
            self.time_steps.append(t)

    def get_history(self):
        """
        Get trajectory history in the format expected by the C++ interface.

        Returns
        -------
        list
            List of trajectory frames, each containing positions for all bodies.
        """
        return [[pos.tolist() for pos in frame] for frame in self.history]

    def get_positions(self):
        """
        Get current positions in the format expected by the C++ interface.

        Returns
        -------
        list
            Current positions as nested list.
        """
        return self.positions.tolist()

    def get_velocities(self):
        """
        Get current velocities in the format expected by the C++ interface.

        Returns
        -------
        list
            Current velocities as nested list.
        """
        return self.velocities.tolist()

    def get_masses(self):
        """
        Get masses in the format expected by the C++ interface.

        Returns
        -------
        list
            Masses as list.
        """
        return self.masses.tolist()

    def get_time_steps(self):
        """
        Get time steps in the format expected by the C++ interface.

        Returns
        -------
        list
            Time steps as list.
        """
        return self.time_steps

    def get_n_bodies(self) -> int:
        """Get number of bodies."""
        return self.n_bodies

    def get_G(self) -> float:
        """Get gravitational constant."""
        return self.G
